javascript-review.js

```diff
@@ -1,90 +1,181 @@
 
+function myConstructor (value) {
+    const obj = {
+        firstVal: value,
+        colour: "red"
+    }
+    return(obj)
+}
+
+console.log(myConstructor(2))
+
 // Note that a constructor function is the old school syntax that was used in javascript before the sugar class syntax came along. 
 
 // 3.
 
 // Can you access variables defined outside of functions within the scope of functions in JS? Show a simple example of this being used? (Define a constant in global scope and use that in a very simple function.)
+//yes
+const globalVar = "WOW"
+
+function myOtherFunc () {
+    console.log(globalVar)
+}
 
+myOtherFunc()
 // 4.
 
 // What two ways can be used to access the properties of objects in JS? Give examples of both (although assume the objects have been defined).
+console.log(myObj["number"])
+console.log(myObj.number)
+//?
 
 // 5.
 
 // What is JSON? Be clear but concise.
+//Javascript Object Notation, a unified syntax for transferring data
 
 // 6.
 
 // What is a callback?
+//a function (reference) that you pass into antoher function (or store in an object?) that can then be used/called in the future
 
 // 7.
 
 // Define a function with that takes two number arguments and a callback. The function will add the two numbers and pass them to the callback as an argument. Now call this function and in the callback simply console.log out the argument that has been passed through.
 
+function one (firstNum, secondNum, callback) {
+    callback(firstNum + secondNum)
+}
+
+function myCallback (value) {
+    console.log(value)
+}
+
+one(2, 3, myCallback)
 // 8.
 
 // Why do we use promises?
+// to handle asyncronous processing (API calls, priority threading etc)
 
 // 9.
 
 // What are the two methods at our disposal if we have hold of a promise? What are the names of the functions that are related to these functions (that are called when the promise is fulfilled or fails).
 
+//then and except
+//? is this the names of the args that you pass in when defining a promise? they can be anything though. If its that then resolve and reject?
+
 // 10.
 
 // What is the DOM?
+// Document object model, an object representation of the HTMl heirarchy of a web page. It can be used by javascript to manipulate HTMl elements
 
 // 11.
 
 // Create an object that has three properties. One property is called ‘count’ and will be set to a number, the second will be set to an array, and the third to a function. This function should simply console.log each of the values of the second property on a new line.
 
+function moreFunc () {
+    this.array.forEach (element => {
+        console.log(element)
+    })
+}
+const moreObj = {
+    count: 3,
+    array: [2, 3, 4],
+    func: moreFunc,
+}
+
+moreObj.func()
 // 12.
 
 // What is NPM? In your brief explanation refer to package.json.
+// node package manager, it lets you install dependencies and stores them in your project package.json (not 100% sure on this one tbh)
```

Comment:
this is correct except that dependencies are stored in the `node_modules` directory